### ISSUE TITLE
What is the current method for applying custom JS modules to custom HTML templates, and loading NPM packages into those modules?

### ISSUE THREAD

**Issue by Mikhail-MM** [bigcommerce/cornerstone#1412]
Can someone please point me to some docs that explain in simple terms how to
1) Install a package with npm
2) Create a custom JS file that can use that package and all other globally installed packages
3) Create a custom HTML template that can be used in any custom route/webpage in the BigCommerce management console
4) Use the previously written JS file in the custom template, and have access to all the NPM packages from the webpack bundle?

When I create a custom JS file, let's say called 'custom-js.js' --- I do not have access to the globally installed packages (So using something like $(window).load gives a "$ is not defined" error, since the JQuery is not in scope)

I am trying to add some custom JavaScript to the Optimized One Page Checkout page on a CornerStone theme.

I've currently found two ways to do this:

Within my checkout.html file, I can paste:

```
<script src="{{cdn 'assets/js/theme/custom-js.js'}}"></script>
 ```

This calls my script, but I can't use any of these:
```
import { throttle } from 'lodash'
import $ from 'jquery'
const lodash = require('lodash')
```
All call an error.

So I believe the CORRECT way to do this is to mount my javascript in APP.js:
```
__webpack_public_path__ = window.__webpack_public_path__; // eslint-disable-line
 
import 'babel-polyfill';
import $ from 'jquery';
import Global from './theme/global';
 
const getAccount = () => import('./theme/account');
const getLogin = () => import('./theme/auth');
const pageClasses = {
...
    checkout: () => import('./theme/custom-js'),
...
};
```
Okay, so this calls my script on the checkout pagebut again, jQuery is undefined, and it is still not connected to the globally imported package definitions. It's like the script is running in isolation outside of the general bundle that powers the rest of cornerstone.

Now I believe these two script tags are needed somewhere to put my file in context:
```
<script src="{{cdn 'assets/dist/theme-bundle.main.js'}}"></script>
<script>
    // Exported in app.js
    window.stencilBootstrap("{{page_type}}", {{jsContext}}).load();
</script>
 ```

Do I need to edit the webpack config file? Please please please help me out with any resources, I would highly appreciate it.

**Comment by Mikhail-MM** [comment]
So after going through the docs some more this is another solution:
Inside of layouts/base.html, we can add this helper:

```
{{inject 'template' template}}
```

And within App.js, we can do this:

```
window.stencilBootstrap = function stencilBootstrap(pageType, contextJSON = null, loadGlobal = true) {
    const context = JSON.parse(contextJSON || {});

    return {
        load() {
            $(async () => {
                let globalClass;
                let pageClass;
                let PageClass;

                console.log(context.template)
                if (context.template === 'pages/my-custom-template') {
                    // Call our custom modules here 
                }
```
But we would be calling our functions here in app.js directly, where they have access to global NPM packages, instead of calling them from their modules. Also, I noticed that  that `console.log(context.template)` was not being called at all in the Optimized One Page Checkout file, so is App.js not firing its context-pageloader functions in that Component? I believe it is an external angular component that is sent to the client, right? 

What are the best options for getting NPM packages working in optimized one page checkout?

**Comment by karen-white** [comment]
Hey @Mikhail-MM , This documentation should help get you most of the way there:
https://developer.bigcommerce.com/stencil-docs/template-files/customize-stencil-checkout/checkout-js-sdk/getting-started-in-vanilla-js

This tutorial creates a new checkout.js file that imports an npm module (the Checkout JS SDK) and then maps the checkout.js file to the checkout.html page. One thing to note is that if you're using the Checkout SDK, you're _replacing_ the existing checkout content. You'll run into a conflict if you're running the changes suggested by the tutorial without commenting out the default checkout content. Are you wanting to augment the existing checkout, or replace the checkout as outlined in the tutorial?

**Comment by Mikhail-MM** [comment]
@karen-white Thank you for the link, I will make sure to take a look. I am mainly looking to augment the existing checkout page, particularly for the default page received by the client when we select *Optimized One Page Checkout*. I had trouble importing Lodash into the context of that page - I don't necessarily want to create an entirely new checkout page, although for a large customization that would be the way to go.

That tutorial walks through creating a custom JS module for an existing page template - I am more generally wondering how to get NPM context into a new page which does not exist yet (And has no mapping at all in app.js)

For example if I went to BC and created a new page in `Storefront > Web Pages`, created a route such as
`/pages/my-custom-page/`, created a new Handlebars Template called `_my-custom-temp.html`, and selected it as the template to be used in the `Storefront > Web Pages` menu.

How can I create a new mapping to that JS in app.js? One idea I had is just to create the custom JS file with various NPM imports and call it within global.js
```
export default class Global extends PageManager {
    onReady() {
        if (context.template === '/pages/my-custom-page/') {
            myNewJs();
        }
    }
}
```
But, I haven't tried this yet.

**Comment by karen-white** [comment]
@Mikhail-MM For mapping a custom JS file to a custom HTML template, @sacr3dc0w submitted a PR awhile back that simplifies the process really nicely. We do need to get our documentation up to date to reflect the changes, though. 

Here's the process summarized in the original issue and the PR that added the necessary updates to Cornerstone:

- Original issue: #1339 

- PR: #1346 

For running your module on the checkout page, have you tried [<script type="module">](https://jakearchibald.com/2017/es-modules-in-browsers/)? I haven't tested this yet, but it might be a way forward.

**Comment by Mikhail-MM** [comment]
@karen-white thank you for the help and the links, I'll check that out, as well as trying the Script type="module" approach.

Are there any plans to open-source the docs for BigCommerce? I think if the docs were open source it would give developers a great opportunity to post some code examples to show how cornerstone can be customized in different ways. One area which I find it very difficult to learn is how to use the custom Handlebars helpers which are custom-built for BC, since there are not too many examples on the docs for each one, and there are not many outside tutorials for BC/Cornerstone specifically

**Comment by karen-white** [comment]
That's good feedback @Mikhail-MM ! We've discussed finding ways to make it easier for external devs to contribute code samples and examples although we don't have any short term plans to make the docs open source. We do have some new resources coming in January that'll include more tutorials, so  keep a lookout!

Agreed on the need for more helpers examples, and I'll pass it on to our docs team. I did want to link you to the [source code](https://github.com/bigcommerce/paper/tree/2.x/helpers) for the Stencil Handlebars helpers in case it's helpful to you. This repo contains the custom helpers, in addition to the built in helpers that we whitelist from [here](https://github.com/helpers/handlebars-helpers).
